### ISSUE TITLE
The Shield setting Block Scripts doesn't block all JS on puredarwin.org

### ISSUE THREAD

**Issue by ghost** [brave/brave-ios#7822]
### Description: 

Script blocking doesn't properly work on puredarwin.org. I have no reason to believe this issue is not widespread. Furthermore, toggling the Shield setting Block Scripts on and off doesn't have any visible effect.

### Steps to Reproduce 
  1. Visit puredarwin.org
  2. Tap the Brave icon in the right bottom corner
  3. Toggle on the Shield setting Block Scripts

### Actual result: 
![539F3890-220B-4CE0-B4BC-D2CA475AAF3D](https://github.com/brave/brave-ios/assets/79327106/f451a809-84bc-487f-8b2f-a8a95a109ee5)

### Expected result:

The YouTube video player embedded into the website should not work.

### Reproduces how often:

Easily reproduced

### Brave Version:

Version 1.52.1 (23.7.11.14)

- Can you reproduce this issue with the most recent build from TestFlight?

I have not attempted to reproduce it with the most recent build from TestFlight.

- Can you reproduce this issue with the previous version of the current build from TestFlight?

I have only ever used builds from AppStore.

- Can you reproduce this issue with the current build from AppStore?

Yes, I can.

### Device details:

iPhone 6s+ (iOS 15.7.8)

### Website problems only:

As in the description, the YouTube video player embedded into the website works when it should not.

### Additional Information

Ever since switching to Brave on all my devices, I noticed a discrepancy between the effect Block Scripts has in both the last build compatible with iOS 12 [Version 1.24 (21.4.1.15)] and, as of writing, the most current one. For example, when Block Scripts is toggled-on and a website offers a noscript variant, it's loaded by the former but not the latter. I don't know whether this is relevant to this issue but I find it concerning that current Brave for iOS is lenient with blocking JavaScript when in my opinion, enabling Block Scripts should block all of it, as expected from its name and concise description. 

### Bonus Question

Is there a technical reason as to why the behaviour between the two aforementioned builds isn't the same? I know about the depreciation of `javaScriptEnabled` in favour of `allowsContentJavaScript` but I don't know what expected technical implications this might've had on JavaScript blocking in Brave if any.

![F721499E-9704-4ADF-96D8-16C691920E30](https://github.com/brave/brave-ios/assets/79327106/3fd8db2b-140d-4931-954a-40bff35c578d)
(puredarwin.org on Brave Version 1.24 (21.4.1.15), Block Scripts is toggled-on, iOS 12)

**Comment by iccub** [comment]
I can confirm this issue does not repro on safari with js toggled off. We are going to look into this

**Comment by Uni-verse** [comment]
Verified on `iPhone X` running `iOS 16.5.1` & `iPad (9th Gen)` running `iPadOS 15.6.1` using version `1.56 (23.8.3.22)`

Using the steps to reproduce in https://github.com/brave/brave-ios/issues/7822#issue-1834052331

- Ensured that embedded video player in `puredarwin.org` is broken/disabled when enabling `Block Scripts` from the brave shields menu.
- Ensured video player is blocked on fresh profile as well as existing profile (1.52.1 -> 1.56).
- Ensured video player is blocked on `puredarwin.org` when enabling `Block Scripts` from the global settings in `Brave Shields and Privacy` and disabling `Block Scripts` per site will enable JavaScript to run. 
- Ensured relaunching the browser app does not change block script setting preference per site
- Ensured per site & global `Block Scripts` will disable JavaScript on other webpages such as `Blizzard.com` (carousel at top of page) and `youtube.com`

1.52.1 | 1.56 | Block Script on other sites  | Global Block Scripts ON
-|-|-|-
![IMG_3438 3](https://github.com/brave/brave-ios/assets/17885425/156d2dc8-1e7c-49e7-b58d-c269e46fe2ea) | ![IMG_3439 2](https://github.com/brave/brave-ios/assets/17885425/3fb67f9e-65dd-4197-8e63-ef333d323e43) | ![IMG_3440 2](https://github.com/brave/brave-ios/assets/17885425/5baca043-ace6-44d6-ae63-474ef69b7cfc) | ![IMG_3441 2](https://github.com/brave/brave-ios/assets/17885425/0096e5b2-5a71-4c67-bb4d-400c6a849548)
![IMG_0320 2](https://github.com/brave/brave-ios/assets/17885425/5f3c36b8-d82f-4e76-a73d-4ab2b87fedb4) | ![IMG_0321 2](https://github.com/brave/brave-ios/assets/17885425/fa58e971-7ad1-4e4d-8f7e-74999e363d4a) | ![IMG_0322 2](https://github.com/brave/brave-ios/assets/17885425/30554fae-7796-4f00-8b75-67b5f1e421f9) | ![IMG_0324 3](https://github.com/brave/brave-ios/assets/17885425/f80a82b0-da01-4a31-83f3-5507431d9eff)

**Comment by mailinglists35** [comment]
@iccub this is actually a general issue, not particular to some site.

In my
experience my conclusion is Javascript OFF does not turn off JS loaded from the same domain, but only third party scripts.

When turning off JS, I expect JS really off unconditionally.

For example this site should display a message about javascript when JS is turned off, and that's exactly how Safari correctly behaves when turning off JS in safari. However in Brave it loads the script but it forbids the script to load remote data from third party URL.

https://amuritiliescu.lol

How Safari treats JS off:

![image](https://github.com/brave/brave-ios/assets/2054302/be4966a1-42e7-44ee-8547-a1c0789e91fd)

How Brave treats JS off:

![image](https://github.com/brave/brave-ios/assets/2054302/5cb23f32-fce2-48de-8f3c-191b370079df)

How the site looks with JS on:

![image](https://github.com/brave/brave-ios/assets/2054302/9217702d-9e14-4a80-9b75-46f076b10f24)

**Comment by mailinglists35** [comment]
this is ofcourse repro on latest Testflight, so no fix landed there about truly turning off JS because there is no fix yet for the behaviour described

**Comment by ghost** [comment]
I cannot verify @mailinglists35's claims, but for what it is worth, Jeff Johnson's in-line JavaScript script tester (https://underpassapp.com/StopTheScript/) does not report that JavaScript code execution is allowed; and, from my own observation, I have always assumed that Brave may simply not communicate to visited websites that JavaScript support is disabled, which might result in the observed behaviour—the user is served with the JavaScript-dependant variant of pages, regardless of the state of the Block Scripts setting.

**Comment by mailinglists35** [comment]
Here is a domain that does not require HTTPS thus you can easily see all plaintext http requests in a proxy setup: 

http://www.cyscape.com/showbrow.asp

with JS off, safari would request a second link http://www.cyscape.com/showbrow.asp?bhjs=0 while Brave would display a blank page and not ask for the second request. 

on desktop, a browser does like Safari, it loads first page then requests the second link.

the link sets a <noscript> meta refresh to the second url, which brave never follows. hm, maybe that's leading me to think it is half working with js off when instead it actually doesn't follow those <noscript> redirects?

**Comment by Brandon-T** [comment]
Hi and thank you for the report and detailed explanation of the issue! :)

I just want to confirm that we are definitely blocking scripts on the URL when `Block Scripts` is enabled. 
It does indeed show a `blank` screen when scripts are blocked as you mentioned in your new ticket:

<img width="836" alt="image" src="https://github.com/brave/brave-ios/assets/1530031/9430e0e4-2f3c-45eb-8b7f-5318bdca20dc">

<img width="465" alt="image" src="https://github.com/brave/brave-ios/assets/1530031/0a115273-022e-4bb4-80cf-be96dbf0a414">

The page has:

```javascript
<noscript>
  <meta http-equiv="refresh" content="0; url='/showbrow.asp?bhjs=0'">
  <a href="/showbrow.asp?bhjs=-1">Please click here to continue</a>
</noscript>
```
Which is should work.

So after some investigation, I can assure you that Javascript is 100% disabled.

I will address your currently open ticket: https://github.com/brave/brave-ios/issues/8426 on this issue.

**EDIT:** In a test application with a plain `WKWebView` and JS disabled, WKWebView does not process the redirect from that meta tag. I thought it was our shields, but that seems to be a different issue.

**Comment by Brandon-T** [comment]
@mailinglists35 

Hey, after further investigation, I realized it's illegal HTML (in WebKit). I just tested:

```html
<html>
  <head>
    <noscript>
      <meta http-equiv="refresh" content="0; url='https://brave.com'">
    </noscript>
    <title>Please wait...</title>
  </head>

  <noscript>
    <body style="background-color: red;">
      <a href="https://brave.com">Please click here to continue</a>
    </body>
  </noscript>
</html>
```
This code will never redirect and never render anything. The `noscript` tag must be inside of a `body` tag.

So:
```html
<html>
  <head>
    <title>Please wait...</title>
  </head>

  <body style="background-color: red;">
    <noscript>
      <a href="https://brave.com">Please click here to continue</a>
    </noscript>
  </body>
</html>
```

**Comment by mailinglists35** [comment]
both html4.01 and html5 allow meta tags in a noscript tag in head tag.

https://www.w3schools.blog/noscript-html

https://html.spec.whatwg.org/multipage/scripting.html#the-noscript-element

**Comment by Brandon-T** [comment]
> both html4.01 and html5 allow meta tags in a noscript tag in head tag.
> 
> https://www.w3schools.blog/noscript-html
> 
> https://html.spec.whatwg.org/multipage/scripting.html#the-noscript-element

Hi you are correct, but WebKit does not obey this (for this issue).
I made a video to show that it isn't just Brave (but also Safari):

https://github.com/brave/brave-ios/assets/1530031/3b67b07a-3f79-4446-af02-84ad607aabbf

and

https://github.com/brave/brave-ios/assets/1530031/149b273c-6ef3-4ecb-b47e-617f592f32a2

The same behaviour happens in Chrome-iOS as well (tested on an actual device for Chrome and Safari). 
As you know, all browsers on iOS are WebKit based at the moment, so we all share the same odd behaviour in this scenario.

As long as `<noscript>` tag is in the `<head>`, it will not work. 
That also means the `<meta>` tag must be inside the `<head>`, and it can't be within the `<noscript>` either.
Additionally, the hyper-link/anchor `<a>` tag must be in the body or within the `<noscript>` in the body (not the `<head>`).

I have tested all possible scenarios to see which works, and it turns out WebKit is ignoring the HTML5 standard here entirely. 

I am filing a `radar` with Apple about this issue and will update here with a link soon.
